backtrace-library/src/main/java/backtraceio/library/metrics/BacktraceEventsHandler.java

```diff
@@ -0,0 +1,202 @@
+package backtraceio.library.metrics;
+
+import android.content.Context;
+import android.os.Handler;
+import android.os.Message;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentLinkedDeque;
+
+import backtraceio.library.common.BacktraceMathHelper;
+import backtraceio.library.common.BacktraceSerializeHelper;
+import backtraceio.library.interfaces.Api;
+import backtraceio.library.logger.BacktraceLogger;
+import backtraceio.library.models.BacktraceMetricsSettings;
+import backtraceio.library.services.BacktraceHandlerThread;
+import backtraceio.library.services.BacktraceReportSender;
+
+public abstract class BacktraceEventsHandler extends Handler {
+
+    private final static transient String LOG_TAG = BacktraceEventsHandler.class.getSimpleName();
+
+    protected final BacktraceHandlerThread backtraceHandlerThread;
+
+    /**
+     * Number of events in the queue
+     */
+    public int getCount() {
+        return events.size();
+    }
+
+    /**
+     * Maximum number of events in store. If number of events in store hit the limit
+     * BacktraceMetrics instance will send data to Backtrace.
+     */
+    private int maxNumEvents = 350;
+
+    public void setMaxNumEvents(int maxNumEvents) {
+        this.maxNumEvents = maxNumEvents;
+    }
+
+    public int getMaxNumEvents() {
+        return this.maxNumEvents;
+    }
+
+    /**
+     * Time between retries if metrics submission fails
+     */
+    private final int timeBetweenRetriesMillis;
+
+    /**
+     * List of events in the event queue
+     */
+    ConcurrentLinkedDeque<Event> events = new ConcurrentLinkedDeque<Event>();
+
+    /**
+     * Number of retry attempts
+     */
+    protected int numRetries = 0;
+
+    /**
+     * Http client
+     */
+    private final Api api;
+
+    /**
+     * Submission url
+     */
+    private final String submissionUrl;
+
+    /**
+     * User provided custom attributes
+     */
+    protected final Map<String, Object> customAttributes;
+
+    /**
+     * The application context
+     */
+    protected final Context context;
+
+    /**
+     * @param context
+     * @param api
+     * @param backtraceHandlerThread
+     * @param urlPrefix
+     * @param settings
+     */
+
+    public BacktraceEventsHandler(Context context, Map<String, Object> customAttributes,
+                                  Api api, final BacktraceHandlerThread backtraceHandlerThread, String urlPrefix, BacktraceMetricsSettings settings)
+    {
+        // This should always have a nonnull looper because BacktraceHandlerThread starts in the
+        // constructor and getLooper blocks until the looper is ready if the thread is started
+        super(backtraceHandlerThread.getLooper());
+        if (backtraceHandlerThread.getLooper() == null) {
```

Comment:
why do we need to invoke it twice?

Reply:
yeah - but why we need to execute it twice? If we already executed getLooper we know it's not null, right?

Reply:
Ah right we can cache it, thank you @konraddysput !

Reply:
@konraddysput we cannot cache it, Java spec requires `super` to be first statement in constructor body so we cannot put a variable declaration and assignment above the `super`.

Reply:
but will there be any situation when getLooper method will return null?

Reply:
![image](https://user-images.githubusercontent.com/9386268/135307250-84234acb-e84b-4197-bda6-0a2988976088.png)